File: cpp/src/solutions/longest_palindromic_substring.cpp
// LeetCode 5. Longest Palindromic Substring by C++
// Manacher Algorithms

#include "common.h"

class Solution {
 public:
  string longestPalindrome(string s) {
    string t = "$#" + string(s.size() * 2, '#');
    for (int i = 0; i < s.size(); i++) {
      t[i * 2 + 2] = s[i];
    }
    vector<int> f(t.size(), 1);
    int current_center = 0, current_right = 0, ans_len = 0, ans_center = 0;
    for (int i = 1; i < t.size(); i++) {
      if (current_right > i) {
        f[i] = min(f[2 * current_center - i], current_right - i);
      }
      while (t[i + f[i]] == t[i - f[i]]) {
        f[i]++;
      }
      if (i + f[i] > current_right) {
        current_right = i + f[i];
        current_center = i;
      }
      if (f[i] > ans_len) {
        ans_len = f[i];
        ans_center = i;
      }
    }
    return s.substr((ans_center - ans_len) / 2, ans_len - 1);
  }
};

TEST(longestPalindrome, example1) {
  string s = "babad", ans = "bab";
  EXPECT_EQ(Solution().longestPalindrome(s), ans);
}

TEST(longestPalindrome, example2) {
  string s = "cbbd", ans = "bb";
  EXPECT_EQ(Solution().longestPalindrome(s), ans);
}

TEST(longestPalindrome, example3) {
  string s = "a", ans = "a";
  EXPECT_EQ(Solution().longestPalindrome(s), ans);
}

TEST(longestPalindrome, example4) {
  string s = "ac", ans = "a";
  EXPECT_EQ(Solution().longestPalindrome(s), ans);
}

File: rust/src/solutions/longest_palindromic_substring.rs
// LeetCode 5. Longest Palindromic Substring in Rust
// Manacher Algorithms

use super::Solution;

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let mut t = vec!['#'; s.len() * 2 + 3];
        *t.first_mut().unwrap()= '$';
        for (i, ch) in s.chars().enumerate() {
            t[2 * i + 2] = ch;
        }
        *t.last_mut().unwrap() = '!';
        let mut f = vec![1; t.len()];
        let (mut current_center, mut current_right, mut ans_len, mut ans_center) = (0, 0, 0, 0);
        for i in 1..t.len() - 1 {
            if current_right > i {
                f[i] = std::cmp::min(f[2 * current_center - i], current_right - i);
            }
            while t[i + f[i]] == t[i - f[i]] {
                f[i] += 1;
            }
            if i + f[i] > current_right {
                current_right = i + f[i];
                current_center = i;
            }
            if f[i] > ans_len {
                ans_len = f[i];
                ans_center = i;
            }
        }
        String::from(&s[(ans_center - ans_len) / 2..(ans_center + ans_len) / 2 - 1])
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s = String::from("babad");
        let ans = String::from("bab");
        assert_eq!(Solution::longest_palindrome(s), ans);
    }

    #[test]
    fn example2() {
        let s = String::from("cbbd");
        let ans = String::from("bb");
        assert_eq!(Solution::longest_palindrome(s), ans);
    }

    #[test]
    fn example3() {
        let s = String::from("a");
        let ans = String::from("a");
        assert_eq!(Solution::longest_palindrome(s), ans);
    }

    #[test]
    fn example4() {
        let s = String::from("ac");
        let ans = String::from("a");
        assert_eq!(Solution::longest_palindrome(s), ans);
    }
}
